context/config_file.go

```diff
@@ -35,32 +31,126 @@ func parseConfigFile(fn string) (*configEntry, error) {
 }
 
 // ParseDefaultConfig reads the configuration file
-func ParseDefaultConfig() (*configEntry, error) {
+func ParseDefaultConfig() (*Config, error) {
 	return parseConfigFile(configFile())
 }
 
-func parseConfig(r io.Reader) (*configEntry, error) {
+type AuthConfig struct {
+	User  string
+	Token string
+}
+
+type HostConfig struct {
+	Host  string
+	Auths []*AuthConfig
+}
+
+type Config struct {
+	Root     *yaml.Node
+	Hosts    []*HostConfig
+	Editor   string
+	Protocol string
+}
+
+func (c *Config) ConfigForHost(hostname string) (*HostConfig, error) {
+	for _, hc := range c.Hosts {
+		if hc.Host == hostname {
+			return hc, nil
+		}
+	}
+	return nil, fmt.Errorf("could not find config entry for \"%s\"", hostname)
+}
+
+func (c *Config) DefaultHostConfig() (*HostConfig, error) {
+	return c.ConfigForHost(defaultHostname)
+}
+
+func defaultConfig() Config {
+	return Config{
+		Protocol: "https",
+		// we leave editor as empty string to signal that we should use environment variables
+	}
+}
+
+func parseConfig(r io.Reader) (*Config, error) {
 	data, err := ioutil.ReadAll(r)
 	if err != nil {
 		return nil, err
 	}
-	var config yaml.Node
-	err = yaml.Unmarshal(data, &config)
+	var root yaml.Node
+	err = yaml.Unmarshal(data, &root)
 	if err != nil {
 		return nil, err
 	}
-	if len(config.Content) < 1 {
+	if len(root.Content) < 1 {
 		return nil, fmt.Errorf("malformed config")
 	}
-	for i := 0; i < len(config.Content[0].Content)-1; i = i + 2 {
-		if config.Content[0].Content[i].Value == defaultHostname {
-			var entries []configEntry
-			err = config.Content[0].Content[i+1].Decode(&entries)
-			if err != nil {
-				return nil, err
+	if root.Content[0].Kind != yaml.MappingNode {
+		return nil, fmt.Errorf("expected a top level map")
+	}
+
+	config := defaultConfig()
+	config.Root = &root
+
+	// TODO
+
+	// - [x] make hosts: nesting format work with new parsing code and config struct
+	// - [ ] support setting editor (or protocol)
+	// - [ ] implement new commands
+	// - [ ] migration code for old (non-hosts) config
+
+	topLevelKeys := root.Content[0].Content
+
+	for i, v := range topLevelKeys {
+		switch v.Value {
+		case "hosts":
+			// hosts is a map of hostname -> arrays of user auth configs
+			for j, v := range topLevelKeys[i+1].Content {
+				if v.Value == "" {
+					continue
+				}
+				hostConfig := HostConfig{}
+				hostConfig.Host = v.Value
+				authsRoot := topLevelKeys[i+1].Content[j+1]
```

Comment:
Is there some sort of guard that this array access won't panic if it's out of bounds?

Reply:
nope -- i still have to add error handling / bounds checking